pkg/image/nodeserver.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package image
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/kubernetes/pkg/util/mount"
+
+	"github.com/kubernetes-csi/drivers/pkg/csi-common"
+)
+
+const (
+	deviceID = "deviceID"
+)
+
+var (
+	TimeoutError = fmt.Errorf("Timeout")
+)
+
+type nodeServer struct {
+	*csicommon.DefaultNodeServer
+	Timeout  time.Duration
+	execPath string
+	args     []string
+}
+
+func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+
+	// Check arguments
+	if req.GetVolumeCapability() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability missing in request")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
+	}
+
+	image := req.GetVolumeContext()["image"]
+
+	err := ns.setupVolume(req.GetVolumeId(), image)
+	if err != nil {
+		return nil, err
+	}
+
+	targetPath := req.GetTargetPath()
+	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(targetPath, 0750); err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			notMnt = true
+		} else {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	if !notMnt {
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	fsType := req.GetVolumeCapability().GetMount().GetFsType()
+
+	deviceId := ""
+	if req.GetPublishContext() != nil {
+		deviceId = req.GetPublishContext()[deviceID]
+	}
+
+	readOnly := req.GetReadonly()
+	volumeId := req.GetVolumeId()
+	attrib := req.GetVolumeContext()
+	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+
+	glog.V(4).Infof("target %v\nfstype %v\ndevice %v\nreadonly %v\nvolumeId %v\nattributes %v\n mountflags %v\n",
+		targetPath, fsType, deviceId, readOnly, volumeId, attrib, mountFlags)
+
+	options := []string{"bind"}
+	if readOnly {
+		options = append(options, "ro")
+	}
+
+	args := []string{"mount", volumeId}
+	ns.execPath = "/bin/buildah" // FIXME
```

Comment:
what is this binary?

Reply:
https://github.com/containers/buildah

Ideally we would get CRI support added for this use case. For now, buildah supports the use case out of the box so useful to bootstrap the project.